lib/SILOptimizer/Mandatory/AddressLowering.h

```diff
@@ -0,0 +1,282 @@
+///
+/// A projection may project out of an operand's definition (def-projection).
+/// After allocation, before materialization or rewriting, we may have:
+///
+///   %result_addr = alloc_stack       // storage for %result
+///   %result = apply : $() -> @out T
+///   %extract = struct_extact %result // def-projection of %result
+///
+/// Or, a projection may project into a composing use (use-projection):
+///
+///   %struct_addr = alloc_stack      // storage for %struct
+///   %result = apply : $() -> @out T // use-projection of %struct at operand #0
+///   %struct = struct %result
+///
+/// A phi-projection is a use projection that projects its entire value
+/// through a phi rather than into a composing use. It has an invalid
+/// 'projectedOperandNum'.
+///
+/// Operations that destructively resuse storage (open_existential_value,
+/// unchecked_enum_data, and switch_enum), are not considered storage
+/// projections. Instead, these values have no ValueStorage but are rewritten to
+/// directly reuse their operand's storage.
+///
+/// To materialize projections, address lowering follows the original def-use
+/// edges for address-only values. Consequently, values that have storage cannot
+/// be removed from SIL or from the storage map until rewriting is
+/// complete. Mapped values can, however, be substituted on-the-fly by emitting
+/// a place-holder value and updating the map entry. This works because the
+/// value storage map holds no direct references to any SIL entities, such as
+/// Operands or SILValues.
+struct ValueStorage {
+  enum : uint32_t { InvalidID = uint32_t(~0) };
+  enum : uint16_t { InvalidOper = uint16_t(~0) };
+
+  /// The final address of this storage after rewriting the SIL. For values
+  /// linked to their own storage, this is set during storage allocation to an
+  /// alloc_stack or indirect function argument. For projections, it is only set
+  /// after materialization (during instruction rewriting).
+  SILValue storageAddress;
+
+  /// When either isDefProjection or isUseProjection is set, this refers to the
+  /// storage whose "def" this value projects out of or whose operand this
+  /// storage projects into via its "use.
+  uint32_t projectedStorageID;
+
+  /// For use-projections, identifies the operand index of the composing use.
+  /// Only valid for non-phi use projections.
+  uint16_t projectedOperandNum;
+
+  /// Projection out of a storage def. e.g. this value is a destructure.
+  unsigned isDefProjection : 1;
+
+  /// Projection into a composing use or phi. e.g. this value is used by a
+  /// struct, tuple, enum, or branch.
+  unsigned isUseProjection : 1;
+
+  // The definition of this value is fully translated to lowered SIL.
+  unsigned isRewritten : 1;
+
+  // This is a use-projection into an enum. Tracked to avoid projecting enums
+  // across phis, which would result in piecewise initialization.
+  unsigned initializesEnum : 1;
+
+  ValueStorage() { clear(); }
+
+  void clear() {
```

Comment:
AFAICT, you are using the `clear` function only in the constructor. Why not just do a member wise initialization?